languages/python/reference/exercise-concepts/robot-simulator.md

```diff
@@ -0,0 +1,62 @@
+
+- higher-order functions:  a function that takes one or more other functions as arguments, _returning_ a function as its retrun value.  The example uses the built-in `property()` as a higher-order function through `@property`.
+
+- `property()` built-in:  a function that returns a property attribute.  When used as a decorator, this transforms the passed-in method into a _getter_ method for read-only attribute with the same name and docstring.
+
+- assignment:  the example uses assignment for all the instance properties and `instructions` dictionary.
+
+- instance properties: This exercise rquires one or more instance properties to persist passed in data.
+
+- mutability: In the example, knowing there are no protected or private properties in python and so conciously mutating `self.x`, `self.y` and `self.compass` through the called instance methods.
+
+- method arguments: the example `__init__` method needs to take `self`, direction, x, and y (coordinates) as arguments.  It also uses `self` and `commands` (a string) for arguments to the `move()` method.
+- default arguments:  Pre-setting a function argument value to protect against it not being passed..  The example uses `direction = NORTH` and `x=0, y=0` to ensure those values for a robot even if they are not passed in initally.
+
+- dictionaries/mapping type: the example uses a dictionary to map paassed in move arguemtns to methods that perform the moves.  The example also uses a dictionary/mapping created by calling `str.maketrans()`.
+
+- bracket notation in context of dictionaries: the example uses passed in move arguments as keys via bracket notation to look up corresponding methods for moving the robot in the _insructions_ dictionary. 
+
+- iteration/iterators/iterating: the example uses a `for loop` to iterate through the passed in command string.
+
+- composition:  adding functionality from a class by incorporating and instance of that class in a class you are creating.  The example cureates a `robot` by instatiating a `compass`, and assiging it to the `self`.compass property of `robot`.
+
+- modulous operator: the example uses the modulus operator to calculate the appropriate _compass_ setting when calling the `right` compass method.
+
+- lists:  the example stores compass direction constants in a `list`
+
+- for loop: the example uses a `for loop` to iterate through the letters of the passed-in `commands` string and looks up the corresponding values in a dictionary, so that the appropriate methods can be called to move the `robot`.
+
+- callable objects:  knowing that appending `()` to the name of an instance method _calls_ it, since instances of classes are _callable_, and implement `__call__()`.
```

Comment:
This is a little problematic as an explanation, since neither class in the example actually implements `__call__`, and indeed most things that are callable _don't_ directly implement it. That could lead to a fair amount of confusion, since `__call__` is usually only implemented when you want to make an instance itself callable, whereas `def` handles implementing it on methods. Also "appending `()` to the name of an instance method _calls_ it` is describing _call semantics_ more than _callable objects_, although I'll admit the way Python handles `object` they're pretty close to the same.

Reply:
Admittedly, I was confused when I wrote that.  I was trying to find a way to describe the examples use of `()` to call the method after the method name was retrieved from the _instruction_ dictionary -- and why you would store the method _name_ without the `()`.  Did not do a good job at it.  Here is a first shot at a re-word, but not really happy with it either:

>  - call semantics:  knowing that appending `()` to the name of an instance method _calls_ it, since instance methods are _callable_.

Reply:
Actually I think that's fine, although maybe provide an example (like `robot.move()` to indicate that the dot-notation needs to be used with methods. Also _might_ be a good idea that arguments _may_ be provided inside the parens.